node/impl/full/eth_trace.go

```diff
@@ -0,0 +1,333 @@
+	// this we need to include the ActorID in the trace. For now, just log a warning and skip
+	// this trace.
+	//
+	// TODO: Add ActorID in trace, see https://github.com/filecoin-project/lotus/pull/11100#discussion_r1302442288
+	from, err := lookupEthAddress(ctx, et.Msg.From, sa)
+	if err != nil {
+		log.Warnf("buildTraces: failed to lookup from address %s: %v", et.Msg.From, err)
+		return nil
+	}
+	to, err := lookupEthAddress(ctx, et.Msg.To, sa)
+	if err != nil {
+		log.Warnf("buildTraces: failed to lookup to address %s: %w", et.Msg.To, err)
+		return nil
+	}
+
+	trace := &ethtypes.EthTrace{
+		Action: ethtypes.EthTraceAction{
+			From:    from.String(),
+			To:      to.String(),
+			Gas:     ethtypes.EthUint64(et.Msg.GasLimit),
+			Input:   nil,
+			Value:   ethtypes.EthBigInt(et.Msg.Value),
+			Method:  et.Msg.Method,
+			CodeCid: et.Msg.CodeCid,
+		},
+		Result: ethtypes.EthTraceResult{
+			GasUsed: ethtypes.EthUint64(et.SumGas().TotalGas),
+			Output:  nil,
+		},
+		Subtraces:    len(et.Subcalls),
+		TraceAddress: addr,
+
+		Parent: parent,
+	}
+
+	trace.SetCallType("call")
+
+	if parent != nil && builtinactors.IsEvmActor(parent.Action.CodeCid) && et.Msg.Method == builtin.MethodsEVM.InvokeContract {
+		log.Debugf("found InvokeContract call at height: %d", height)
+		trace.Action.Input, err = decodePayload(et.Msg.Params, et.Msg.ParamsCodec)
+		if err != nil {
+			return xerrors.Errorf("buildTraces: %w", err)
+		}
+		trace.Result.Output, err = decodePayload(et.MsgRct.Return, et.MsgRct.ReturnCodec)
+		if err != nil {
+			return xerrors.Errorf("buildTraces: %w", err)
+		}
+	} else if et.Msg.To == builtin.EthereumAddressManagerActorAddr &&
+		et.Msg.Method == builtin.MethodsEAM.CreateExternal {
+		log.Debugf("found CreateExternal call at height: %d", height)
+		trace.Action.Input, err = decodePayload(et.Msg.Params, et.Msg.ParamsCodec)
+		if err != nil {
+			return xerrors.Errorf("buildTraces: %w", err)
+		}
+
+		if et.MsgRct.ExitCode.IsSuccess() {
+			// ignore return value
+			trace.Result.Output = nil
+		} else {
+			// return value is the error message
+			trace.Result.Output, err = decodePayload(et.MsgRct.Return, et.MsgRct.ReturnCodec)
+			if err != nil {
+				return xerrors.Errorf("buildTraces: %w", err)
+			}
+		}
+
+		// treat this as a contract creation
+		trace.SetCallType("create")
+	} else {
+		trace.Action.Input, err = handleFilecoinMethodInput(et.Msg.Method, et.Msg.ParamsCodec, et.Msg.Params)
+		if err != nil {
+			return xerrors.Errorf("buildTraces: %w", err)
+		}
+		trace.Result.Output, err = handleFilecoinMethodOutput(et.MsgRct.ExitCode, et.MsgRct.ReturnCodec, et.MsgRct.Return)
+		if err != nil {
+			return xerrors.Errorf("buildTraces: %w", err)
+		}
+	}
+
+	// TODO: is it OK to check this here or is this only specific to certain edge case (evm to evm)?
+	if et.Msg.ReadOnly {
+		trace.SetCallType("staticcall")
+	}
+
+	// there are several edge cases thar require special handling when displaying the traces
+	if parent != nil {
```

Comment:
I'm curious why you're looking backwards instead of forwards. I.e., the other way to write this is:

1. Detect calls to the `Exec` method on the init actor.
2. Then run that sub-trace through a special method to find the call to the constructor.

Reply:
I'd at least explicitly state that we're "editing" the trace backwards.